### PR TITLE
feat: add sms provider sendblue

### PR DESCRIPTION
**Pull request by jeffreyhugh**
## What kind of change does this PR introduce?

Feature

## Additional context

SendBlue is an iMessage API. It's under development, but it's actively developed, has a paying userbase, works well enough to send a one-time passcode.

I tested SendBlue by sending an OTP to myself through GoTrue and it works. ~~At the moment, there are no unit tests because I couldn't get past the `gock: cannot match any request` error, but I'll keep poking at it, and hopefully I can get some committed soon. I'm putting the PR out now so I can iterate on feedback while trying to fix the unit test issues.~~

In the meantime, the test (?) keys on the [SendBlue docs](https://sendblue.co/docs) seem to work for trying these changes.

EDIT: SendBlue is under development

### CODE REVIEW

**Comment by jeffreyhugh** [comment]
The latest commit added unit tests. I think this is ready to merge; let me know what the team thinks.

**Comment by jeffreyhugh** [comment]
Any blockers on this PR?

**Comment by Thore1954** [comment]
When #558 is merged, anyone can use any sms provider like this second-tier degraded one.

File: api/sms_provider/sendblue.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+type SendBlueProvider struct {
+	Config  *conf.SendBlueProviderConfiguration
+	APIPath string
+}
+
+type SendBlueResponse struct {
+	Message      string `json:"message"` // this field is for auth-related errors (e.g. bad token)
+	ErrorCode    int    `json:"error_code"`
+	ErrorMessage string `json:"error_message"` // this field is for message-related errors (e.g. blacklisted number), https://sendblue.co/docs/outbound#error-codes
```
Comment (hf) [inline api/sms_provider/sendblue.go:24]:
```suggestion
	ErrorMessage string `json:"error_message"` // this field is for message-related errors (e.g. blocked number), https://sendblue.co/docs/outbound#error-codes
```

File: api/sms_provider/sendblue.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	}, nil
+}
+
+// Send an SMS containing the OTP with SendBlue's API
+// See: https://sendblue.co/docs/outbound
+func (t *SendBlueProvider) SendSms(phone string, message string) error {
+	body, err := json.Marshal(map[string]string{
+		"number":  "+" + phone, // Tests suggest SendBlue does not require a '+', but the docs have a '+'
+		"content": message,
+	})
```
Comment (hf) [inline api/sms_provider/sendblue.go:46]: Please use a Go struct instead.

File: README.md [hunk 3]
```diff
@@ -705,6 +705,11 @@ Or Messagebird credentials, which can be obtained in the [Dashboard](https://das
 - `SMS_MESSAGEBIRD_ACCESS_KEY` - your Messagebird access key
 - `SMS_MESSAGEBIRD_ORIGINATOR` - SMS sender (your Messagebird phone number with + or company name)
 
+Or SendBlue credentials, available at the [Dashboard](https://sendblue.co/dashboard/api)
+
+- `SMS_SENDBLUE_KEY_ID` - called "API Key" (first field) on the dashboard
+- `SMS_SENDBLUE_SECRET_KEY` - called "API Secret" (second field) on the dashboard
```
Comment (hf) [inline README.md:711]: Can you call the config variables the same as they're called in the SendBlue API?

So `SMS_SENDBLUE_API_KEY` and `SMS_SENDBLUE_API_SECRET`. This should avoid any confusion.

File: api/sms_provider/sendblue.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{Timeout: defaultTimeout}
+	r, err := http.NewRequest("POST", t.APIPath, bytes.NewBuffer(body))
```
Comment (hf) [inline api/sms_provider/sendblue.go:52]: Use `http.MethodPost` instead of `"POST"`.

**Comment by kangmingtay** [comment]
Hey @jeffreyhugh, thanks for the contribution! Unfortunately, i don't think we plan to merge this in since the team has decided to go with webhooks in the future for custom sms providers which would be more scalable in terms of maintaining the codebase. We've left some comments on your PR in case you want to use it in a self-hosted setup. Really appreciate you taking the time to make this PR!

**Pull request closed without merging**